spec/consensus/reactor/reactor.md

```diff
@@ -0,0 +1,327 @@
+This distributed data-structure is easily described as a set per process in which elements are added to include them G-Set; the sets kept by processes are approximations of the G-Set.
+
+The [2 Phase-Set](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#2P-Set_(Two-Phase_Set)) (2P-Set) is a variation that allows removals.
+It combines two sets, one in which elements are added to include them in the 2P-Set, $A$, and one in which elements are added to remove them from the 2P-Set, $D$; the actual membership of the 2P-Set is given by $A \setminus D$.
+
+
+
+
+> Warning/TODO: A word about tombstones, that is $D$
+> - Only state that is not required should be deleted/tombstone'd.
+> - Instead of tombstones, add new entries that trigger removal of other entries (for example, state about a new height); each node must be given information to realize by itself that an entry is no longer needed.
+> - Tombstones are an optimization, kept to prevent data recreation and redeletion.
+> - Tombstones should be garbage collected at some point; imprecision shouldn't affect correctness/termination, since this is an optimization (as long as deleted state is never required again).
+> - Tombstones are not to be gossiped; if they were, they would need to carry proof for the reason they were created, defeating their point.
+
+
+
+
+
+# Part 2: CONS/GOSSIP interaction
+
+CONS, the Consensus Reactor State Layer, is where the actions of the Tendermint algorithm are implemented.
+Actions are executed once certain pre-conditions apply, such as timeout expirations or reception of information from validators in the system, neighbors or not.
+
+An action may require communicating with applications and other reactors, for example to gather data to compose a proposal or to deliver decisions, and with the P2P layer, to communicate with other nodes.
+
+## Northbound Interaction - ABCI
+
+This specification focuses on the southbound interactions of CONS, with the GOSSIP and through GOSSIP with P2P.
+
+For those interested in the interactions of CONS with applications and other reactors, we redirect the readers to the [Application Blockchain Interface (ABCI)](../../abci/) specification, which covers most of such communication.
+ABCI specifies both what CONS [requires from the applications](../../abci/abci%2B%2B_app_requirements.md) and on what CONS [provides to Applications](../../abci/abci%2B%2B_tmint_expected_behavior.md).
+
+Interactions with other reactors, such as with the Mempool reactor to build tentative proposals, will be covered elsewhere.
+
+
+
+## Southbound Interaction - GOSSIP-I
+CONS interacts southbound only with GOSSIP, to update the gossip state and to evaluate the current state to check for conditions that enable actions.
+
+To update the state, CONS passes a tuple GOSSIP a tuple with the exact content to be added to the tuple space through a functional call.
```

Comment:
It's not obvious where the tuple space lives. I would have expected it to live inside the CONS layer intuitively. But here it's implied that GOSSIP holds the tuple space. Might be good to clarify which layer holds $t_p$.